testsuite/cluster-test/src/atomic_histogram.rs

```diff
@@ -0,0 +1,232 @@
+const DEFAULT_HISTOGRAM_STEP_MILLIS: u64 = 50;
+
+pub struct AtomicHistogramAccumulator {
+    capacity: usize,
+    step_millis: u64,
+    buckets: Vec<AtomicU64>,
+}
+
+impl Default for AtomicHistogramAccumulator {
+    fn default() -> AtomicHistogramAccumulator {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_MILLIS)
+    }
+}
+
+impl AtomicHistogramAccumulator {
+    pub fn new(size: usize, step: u64) -> AtomicHistogramAccumulator {
+        let mut buf = Vec::with_capacity(size);
+        for _i in 0..size {
+            buf.push(AtomicU64::new(0));
+        }
+        Self {
+            capacity: size,
+            step_millis: step,
+            buckets: buf,
+        }
+    }
+
+    pub fn snapshot(&self) -> AtomicHistogramSnapshot {
+        let mut buf = Vec::with_capacity(self.capacity);
+        for i in 0..self.capacity {
+            buf.push(self.buckets[i].load(Ordering::Relaxed));
+        }
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_millis: self.step_millis,
+            buckets: buf,
+        }
+    }
+
+    fn get_bucket_num(&self, latency: u64) -> u64 {
+        let bucket_num = latency / self.step_millis;
+        if bucket_num >= self.capacity as u64 - 2 {
+            return self.capacity as u64 - 1;
+        }
+        bucket_num
+    }
+
+    pub fn record_latency_ms(&self, latency_ms: u64, num_requests: u64) {
+        let bucket_num = self.get_bucket_num(latency_ms);
+        self.buckets[bucket_num as usize].fetch_add(num_requests as u64, Ordering::Relaxed);
+    }
+}
+
+pub struct AtomicHistogramSnapshot {
+    capacity: usize,
+    step_millis: u64,
+    buckets: Vec<u64>,
+}
+
+impl Default for AtomicHistogramSnapshot {
+    fn default() -> AtomicHistogramSnapshot {
+        AtomicHistogramAccumulator::new(DEFAULT_HISTOGRAM_CAPACITY, DEFAULT_HISTOGRAM_STEP_MILLIS)
+            .snapshot()
+    }
+}
+
+// reserving place for debug print in the future
+impl Debug for AtomicHistogramSnapshot {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "",)
+    }
+}
+
+pub fn sub_vec(a: &[u64], b: &[u64]) -> Vec<u64> {
+    let mut buf = Vec::with_capacity(a.len());
+    if a.len() != b.len() {
+        for &i in a.iter() {
+            buf.push(i);
+        }
+    } else {
+        for i in 0..a.len() {
+            buf.push(a[i] - b[i]);
+        }
+    }
+    buf
+}
+
+impl Sub for &AtomicHistogramSnapshot {
+    type Output = AtomicHistogramSnapshot;
+
+    fn sub(self, other: &AtomicHistogramSnapshot) -> AtomicHistogramSnapshot {
+        AtomicHistogramSnapshot {
+            capacity: self.capacity,
+            step_millis: self.step_millis,
+            buckets: sub_vec(self.buckets.as_ref(), other.buckets.as_ref()),
+        }
+    }
+}
+
+impl AtomicHistogramSnapshot {
+    pub fn cal_percentile_latency(&self, numerator: u64, denominator: u64, stats: &TxStats) -> u64 {
```

Comment:
We should not pass TxStats here. Please make sure this file does not use anything from tx_emitter